core/src/main/java/io/questdb/cairo/sql/async/PageFrameSequence.java

```diff
@@ -0,0 +1,439 @@
+
+        // now we need to steal all dispatch0 work from the queue until we reach either:
+        // - end of the queue
+        // - find out publisher id
+        final MCSequence dispatchSubSeq = messageBus.getPageFrameDispatchSubSeq();
+        while (true) {
+            final long c = dispatchSubSeq.next();
+            if (c > -1) {
+                PageFrameSequence<?> s = pageFrameDispatchQueue.get(c).getFrameSequence();
+
+                if (s == this) {
+                    dispatchSubSeq.done(c);
+                    break;
+                }
+
+                // We are entering all frame sequences in work stealing mode, which means
+                // they can end up being partially dispatched. This is done because we
+                // cannot afford to enter infinite loop here
+                PageFrameDispatchJob.handleTask(s, rec, messageBus, true, circuitBreaker);
+            } else {
+                break;
+            }
+        }
+    }
+
+    /**
+     * Prepares page frame sequence for retrieving the same data set again. The method
+     * is not thread-safe.
+     */
+    public void toTop() {
+        if (frameCount > 0) {
+            LOG.info().$("toTop [shard=").$(shard)
+                    .$(", id=").$(id)
+                    .I$();
+
+            await();
+
+            // done latch is reset by method call above
+            doneLatch.reset();
+            dispatchStartIndex.set(0);
+            reduceCounter.set(0);
+            long dispatchCursor;
+            do {
+                dispatchCursor = dispatchPubSeq.next();
+                if (dispatchCursor < 0) {
+                    stealWork();
+                } else {
+                    break;
+                }
+            } while (true);
+            final PageFrameDispatchTask dispatchTask = pageFrameDispatchQueue.get(dispatchCursor);
+            dispatchTask.of(this);
+            dispatchPubSeq.done(dispatchCursor);
+        }
+    }
+
+    private static int getWorkerId() {
+        final Thread thread = Thread.currentThread();
+        final int workerId;
+        if (thread instanceof Worker) {
+            workerId = ((Worker) thread).getWorkerId() + 1;
+        } else {
+            workerId = 0;
+        }
+        return workerId;
+    }
+
+    private void dispatch0() {
+        // We need to subscribe publisher sequence before we return
+        // control to the caller of this method. However, this sequence
+        // will be unsubscribed asynchronously.
+        messageBus.getPageFrameCollectFanOut(shard).and(collectSubSeq);
+        LOG.debug()
+                .$("added [shard=").$(shard)
+                .$(", id=").$(id)
+                .$(", seqCurrent=").$(collectSubSeq.current())
+                .I$();
+
+        long dispatchCursor;
+        do {
+            dispatchCursor = dispatchPubSeq.next();
+
+            if (dispatchCursor < 0) {
+                stealWork();
+                LockSupport.parkNanos(1);
```

Comment:
Should be `Os.pause()`

Reply:
fixed https://github.com/questdb/questdb/pull/1732/commits/edcb15659325bda39ad82b2f68e46b5dac3d2dd8